### DISCUSSION TITLE
Is the performance better than MST or same?

### DISCUSSION THREAD

**Discussion by swayf** [xaviergonz/mobx-keystone#245] [Q&A]
First of all, thank you, for very interesting library. It looks awesome! =) (MST is very nice too =))

Have you done some performance tests and comparison with MST?

**Comment by xaviergonz** [comment]
Depends, I made some basic benchmarks (as of v0.12.1)

1. When creating models MST is faster since it uses lazy initialization and mobx-keystone uses eager initialization.

2. When creating models and accessing all variables (which forces MST to create all nodes) their speed is roughly equivalent.

3. Reading variables if roughly equivalent.

4. Writing to props is 2x-3x faster in mobx-keystone. 

Benchmarks were done with the production version of both libs and runtime type-checking disabled for mobx-keystone (since MST turns off checking in production mode)

```
(1) empty creation (type checking disabled) - mst is usually faster here due to lazy creation
mobx-keystone x 3,335 ops/sec ±5.03% (80 runs sampled)
mobx-state-tree x 8,766 ops/sec ±1.56% (87 runs sampled)
Fastest is mobx-state-tree by 2.63x

(2) empty creation + access all simple props (type checking disabled)
mobx-keystone x 3,505 ops/sec ±1.97% (82 runs sampled)
mobx-state-tree x 3,384 ops/sec ±2.73% (86 runs sampled)
Fastest is mobx-keystone by 1.04x

(3) already created, access all simple props (type checking disabled)
mobx-keystone x 325,223 ops/sec ±1.54% (90 runs sampled)
mobx-state-tree x 329,137 ops/sec ±0.71% (91 runs sampled)
Fastest is mobx-state-tree,mobx-keystone by 1.01x

(4) already created, change all simple props (type checking disabled)
mobx-keystone x 6,060 ops/sec ±6.34% (91 runs sampled)
mobx-state-tree x 2,550 ops/sec ±0.84% (90 runs sampled)
Fastest is mobx-keystone by 2.38x

(4) already created, change one simple props + getSnapshot (type checking disabled)
mobx-keystone x 173,620 ops/sec ±1.82% (91 runs sampled)
mobx-state-tree x 60,499 ops/sec ±4.91% (87 runs sampled)
Fastest is mobx-keystone by 2.87x
```

**Comment by terrysahaidak** [comment]
I've tried it again on my machine with 0.14.0:

```
empty creation (type checking disabled) - mst is usually faster here due to lazy creation
mobx-keystone x 4,885 ops/sec ±2.70% (86 runs sampled)
mobx-state-tree x 8,928 ops/sec ±4.76% (88 runs sampled)
Fastest is mobx-state-tree by 1.83x

empty creation + access all simple props (type checking disabled)
mobx-keystone x 4,654 ops/sec ±1.28% (85 runs sampled)
mobx-state-tree x 3,474 ops/sec ±1.15% (89 runs sampled)
Fastest is mobx-keystone by 1.34x

already created, access all simple props (type checking disabled)
mobx-keystone x 401,635 ops/sec ±0.32% (96 runs sampled)
mobx-state-tree x 335,645 ops/sec ±0.42% (96 runs sampled)
Fastest is mobx-keystone by 1.20x

already created, change all simple props (type checking disabled)
mobx-keystone x 6,681 ops/sec ±0.65% (91 runs sampled)
mobx-state-tree x 2,410 ops/sec ±0.55% (93 runs sampled)
Fastest is mobx-keystone by 2.77x

already created, change one simple props + getSnapshot (type checking disabled)
mobx-keystone x 183,698 ops/sec ±1.31% (93 runs sampled)
mobx-state-tree x 59,848 ops/sec ±2.00% (89 runs sampled)
Fastest is mobx-keystone by 3.07x
```

**Comment by jrmyio** [comment]
@xaviergonz can you explain where MST and this keystone differ? 

In the readme you never mention MST. I bet you have used MST before starting this library. Would love to know your reasoning of creating your own.

Personally, I love MST but my biggest issue with it is the slowness of creating/accessing the tree on the initial load (in comparison to plain mobx). This is most look likely because it copies many actions / functions for each individual node rather than relying on javascripts' prototype.

Are there any optimizations possible in keystone that could make it significantly faster then MST in this area? I guess OP asking about performance in relation to MST has similar issues with MST.

**Comment by xaviergonz** [comment]
@ConneXNL 
Yes, I have used MST, actually I am a mantainer there, though I've been less active there because of this project

As for differences, they are on the docs site:
https://mobx-keystone.js.org/mstComparison

and also here
https://medium.com/@xaviergonz/mobx-keystone-an-alternative-to-mobx-state-tree-without-some-of-its-pains-8140767a3aa1

mobx-keystone does use prototype for actions / functions, however it does not have lazy initialization.

If you have a particular example made with MST I could port it to mobx-keystone and benchmark both.

**Comment by terrysahaidak** [comment]
Can confirm the performance boost of model creation in `0.21.0` by #36:

```
empty creation (type checking disabled) - mst is usually faster here due to lazy creation
mobx-keystone x 5,211 ops/sec ±2.41% (88 runs sampled)
mobx-state-tree x 8,863 ops/sec ±4.10% (86 runs sampled)
Fastest is mobx-state-tree by 1.70x

empty creation + access all simple props (type checking disabled)
mobx-keystone x 4,782 ops/sec ±1.15% (93 runs sampled)
mobx-state-tree x 3,506 ops/sec ±0.99% (90 runs sampled)
Fastest is mobx-keystone by 1.36x

already created, access all simple props (type checking disabled)
mobx-keystone x 388,299 ops/sec ±0.38% (94 runs sampled)
mobx-state-tree x 337,817 ops/sec ±0.34% (96 runs sampled)
Fastest is mobx-keystone by 1.15x

already created, change all simple props (type checking disabled)
mobx-keystone x 6,216 ops/sec ±3.11% (81 runs sampled)
mobx-state-tree x 2,247 ops/sec ±4.16% (87 runs sampled)
Fastest is mobx-keystone by 2.77x

already created, change one simple props + getSnapshot (type checking disabled)
mobx-keystone x 170,720 ops/sec ±1.47% (84 runs sampled)
mobx-state-tree x 60,202 ops/sec ±2.09% (89 runs sampled)
Fastest is mobx-keystone by 2.84x
```
